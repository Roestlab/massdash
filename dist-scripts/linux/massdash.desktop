[Desktop Entry]
Version=0.0.8
Type=Application
Name=massdash
Exec=/usr/local/bin/massdash
Icon=/usr/share/icons/massdash.ico
Terminal=true
