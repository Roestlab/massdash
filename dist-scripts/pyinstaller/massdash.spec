# -*- mode: python ; coding: utf-8 -*-

import site
import importlib.metadata
import os
import sys

from PyInstaller.building.build_main import Analysis, PYZ, EXE, COLLECT
from PyInstaller.utils.hooks import collect_all, collect_data_files, copy_metadata

from transformers.dependency_versions_check import pkgs_to_check_at_runtime

sys.setrecursionlimit(sys.getrecursionlimit() * 5)

##################### User definitions
exe_name = 'massdash_gui'
script_name = 'run.py'
if sys.platform[:6] == "darwin":
	icon = '../../massdash/assets/img/MassDash_Logo.icns'
else:
	icon = '../../massdash/assets/img/MassDash_Logo.ico'
block_cipher = None
location = os.getcwd()
project = "massdash"
remove_tests = True
bundle_name = "massdash"
#####################


requirements = {project, "streamlit_javascript", "upsetplot", "distributed"}
datas = [(f"{site.getsitepackages()[0]}/streamlit/runtime", "./streamlit/runtime")]
hidden_imports = set()
binaries = []
checked = set()
while requirements:
	requirement = requirements.pop()
	print(f"Info: Checking {requirement}")
	checked.add(requirement)
	if requirement in ["pywin32"]:
		continue
	try:
		module_version = importlib.metadata.version(requirement)
	except (
		importlib.metadata.PackageNotFoundError,
		ModuleNotFoundError,
		ImportError
	):
		continue
	try:
		datas_, binaries_, hidden_imports_ = collect_all(
			requirement,
			include_py_files=True
		)
	except ImportError:
		continue
	datas += datas_
	# binaries += binaries_
	hidden_imports_ = set(hidden_imports_)
	if "" in hidden_imports_:
		hidden_imports_.remove("")
	if None in hidden_imports_:
		hidden_imports_.remove(None)
	requirements |= hidden_imports_ - checked
	hidden_imports |= hidden_imports_

if remove_tests:
	hidden_imports = sorted(
		[h for h in hidden_imports if "test" not in h.split(".")]
	)
else:
	hidden_imports = sorted(hidden_imports)


hidden_imports = [h for h in hidden_imports if "__pycache__" not in h]
datas = [d for d in datas if ("__pycache__" not in d[0]) and (d[1] not in [".", "Resources", "scripts"])]

#if sys.platform[:5] == "win32":
#	base_path = os.path.dirname(sys.executable)
#	library_path = os.path.join(base_path, "Library", "bin")
#	dll_path = os.path.join(base_path, "DLLs")
#	libcrypto_dll_path = os.path.join(dll_path, "libcrypto-1_1-x64.dll")
#	libssl_dll_path = os.path.join(dll_path, "libssl-1_1-x64.dll")
#	libcrypto_lib_path = os.path.join(library_path, "libcrypto-1_1-x64.dll")
#	libssl_lib_path = os.path.join(library_path, "libssl-1_1-x64.dll")
#	if not os.path.exists(libcrypto_dll_path):
#		datas.append((libcrypto_lib_path, "."))
#	if not os.path.exists(libssl_dll_path):
#		datas.append((libssl_lib_path, "."))

for _pkg in ["python","accelerate"]:
	if _pkg in pkgs_to_check_at_runtime:
		pkgs_to_check_at_runtime.remove(_pkg)
for _pkg in pkgs_to_check_at_runtime:
	datas += copy_metadata(_pkg)


datas += collect_data_files("streamlit")
datas += copy_metadata("streamlit")
datas += collect_data_files("pyopenms")
datas += copy_metadata("pyopenms")
datas += collect_data_files("massdash")
datas += copy_metadata("massdash")

hidden_imports = ['pyopenms', 'massdash', 'torchaudio.lib.libtorchaudio']

a = Analysis(
	[script_name],
	pathex=[location],
	binaries=binaries,
	datas=datas,
	hiddenimports=hidden_imports,
	hookspath=['./hooks'],
	runtime_hooks=[],
	excludes=[],
	win_no_prefer_redirects=False,
	win_private_assemblies=False,
	cipher=block_cipher,
	noarchive=False
)
pyz = PYZ(
	a.pure,
	a.zipped_data,
	cipher=block_cipher
)

if sys.platform[:5] == "linux":
	exe = EXE(
		pyz,
		a.scripts,
		a.binaries,
		a.zipfiles,
		a.datas,
		name=bundle_name,
		debug=False,
		bootloader_ignore_signals=False,
		strip=False,
		upx=True,
		console=True,
		upx_exclude=[],
		icon=icon
	)
else:
	exe = EXE(
		pyz,
		a.scripts,
		# a.binaries,
		a.zipfiles,
		# a.datas,
		exclude_binaries=True,
		name=exe_name,
		debug=False,
		bootloader_ignore_signals=False,
		strip=False,
		upx=True,
		console=True,
		icon=icon
	)
	coll = COLLECT(
		exe,
		a.binaries,
		# a.zipfiles,
		a.datas,
		strip=False,
		upx=True,
		upx_exclude=[],
		name=exe_name
	)