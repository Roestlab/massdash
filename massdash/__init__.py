"""
massdash
~~~~~~~~
TODO: ADD general documentation about how to use the auto API here

Package Structure
-----------------
testing - a test package
"""

__all__ = [ 'structs', 'preprocess', 'loaders', 'ui', 'server', 'testing' ]
__version__ = "0.0.9"
