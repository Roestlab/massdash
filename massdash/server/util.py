"""
massdash/server/util
~~~~~~~~~~~~~~~~~~~~
"""

import mmap

def get_string_mslevels_from_bool(mslevel_bool_dict):
    """
    Converts a dictionary of boolean values for mslevels to a string.

    Args:
    mslevel_bool_dict : dict
        A dictionary containing boolean values for mslevels.

    Returns:
    mslevel_str : str
        A string representing the selected mslevel.
    """
    if mslevel_bool_dict['ms1']:
        mslevel_str = 'ms1'
    elif mslevel_bool_dict['ms2']:
        mslevel_str = 'ms2'
    elif mslevel_bool_dict['ms1'] and mslevel_bool_dict['ms2']:
        mslevel_str = 'ms1ms2'
    else:
        raise ValueError('No mslevel selected')
    return mslevel_str

def check_ion_mobility(mzml_file, num_lines_to_check=10_000_000):
    """
    Check if the mzML file contains ion mobility data

    Args:
    mzml_file: (str) mzML file to load
    num_lines_to_check: (int) Number of lines to check for "Ion Mobility"

    Returns:
    Return a boolean indicating if the mzML file contains ion mobility data
    """
    has_ion_mobility = False
    with open(mzml_file, 'rb', 0) as file:
        s = mmap.mmap(file.fileno(), 0, access=mmap.ACCESS_READ) 
        for line in iter(s.readline, b""):
            if b'Ion Mobility' in line:
                has_ion_mobility = True
                break 
            if s.tell() > num_lines_to_check:
                break
    return has_ion_mobility