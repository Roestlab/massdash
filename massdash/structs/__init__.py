"""
massdash/structs
~~~~~~~~~~~~~~~~

Internal Data Structures for storing data 
"""

from .Chromatogram import Chromatogram
from .Data1D import Data1D
from .FeatureMap import FeatureMap
from .FeatureMapCollection import FeatureMapCollection
from .GenericFeature import GenericFeature
from .GenericStructCollection import GenericStructCollection
from .Mobilogram import Mobilogram
from .Spectrum import Spectrum
from .TargetedDIAConfig import TargetedDIAConfig
from .TopTransitionGroupFeatureCollection import TopTransitionGroupFeatureCollection
from .TransitionFeature import TransitionFeature
from .TransitionGroupFeature import TransitionGroupFeature
from .TransitionGroup import TransitionGroup
from .TransitionGroupCollection import TransitionGroupCollection
from .TransitionGroupFeatureCollection import TransitionGroupFeatureCollection


__all__ = [ 
            "Chromatogram", 
            "Data1D",
            "FeatureMap", 
            "FeatureMapCollection",
            "GenericFeature",
            "GenericStructCollection",
            "Mobilogram", 
            "TargetedDIAConfig",
            "TransitionFeature", 
            "TransitionGroup", 
            "TopTransitionGroupFeatureCollection",
            "TransitionGroupCollection",
            "TransitionGroupFeature", 
            "TransitionGroupFeatureCollection",
            "Spectrum"]
