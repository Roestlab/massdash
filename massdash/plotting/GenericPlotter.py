"""
massdash/plotting/GenericPlotter
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
"""


from abc import ABC, abstractmethod
from typing import List, Optional, Literal

# Structs
from ..structs.TransitionGroupFeature import TransitionGroupFeature
from ..structs.TransitionGroup import TransitionGroup

class PlotConfig:
    """
    A class representing the configuration for a plot.

    Attributes:
    -----------
    title : str or None
        The title of the plot.
    subtitle : str or None
        The subtitle of the plot.
    x_axis_label : str
        The label for the x-axis.
    y_axis_label : str
        The label for the y-axis.
    smoothing_dict : dict
        A dictionary containing the parameters for smoothing the data.
    x_range : tuple or None
        The range of values to be displayed on the x-axis.
    y_range : tuple or None
        The range of values to be displayed on the y-axis.
    scale_intensity : bool
        A flag indicating whether to scale the intensity of the data.
    """
    def __init__(self):
        self.include_ms1 = True
        self.include_ms2 = True
        self.num_plot_columns = 2
        self.title = None
        self.subtitle = None
        self.x_axis_label = "Retention Time"
        self.y_axis_label = "Intensity"
        self.smoothing_dict = {'type': 'sgolay', 'sgolay_polynomial_order': 3, 'sgolay_frame_length': 11}
        self.x_range = None
        self.y_range = None
        self.scale_intensity = False
        self.aggregate_mslevels = False
        self.type_of_heatmap = "m/z vs retention time"
        self.type_of_3d_plot = "3D Scatter Plot"
        self.type_of_comparison = "retention time vs ion mobility"
        self.context = "streamlit" # or "jupyter"

    def __str__(self):
        return f"{'-'*8} PlotConfig {'-'*8}\ninclude_ms1: {self.include_ms1}\ninclude_ms2: {self.include_ms2}\nnum_plot_columns: {self.num_plot_columns}\ntitle: {self.title}\nsubtitle: {self.subtitle}\nx_axis_label: {self.x_axis_label}\ny_axis_label: {self.y_axis_label}\nsmoothing_dict: {self.smoothing_dict}\nx_range: {self.x_range}\ny_range: {self.y_range}\nscale_intensity: {self.scale_intensity}\naggregate_mslevels: {self.aggregate_mslevels}\ntype_of_heatmap: {self.type_of_heatmap}\ntype_of_3d_plot: {self.type_of_3d_plot}\ntype_of_comparison: {self.type_of_comparison}\n{'-'*30}"

    def update(self, config_dict):
        """
        Update the configuration attributes using a dictionary.

        Args:
            config_dict (dict): A dictionary containing configuration values.
        """
        for key, value in config_dict.items():
            if hasattr(self, key):
                setattr(self, key, value)

class GenericPlotter(ABC):
    """ 
    This is a generic plotter class 
    """

    def __init__(self, config: PlotConfig):
        self.include_ms1 = config.include_ms1
        self.include_ms2 = config.include_ms2
        self.title = config.title
        self.subtitle = config.subtitle
        self.x_axis_label = config.x_axis_label
        self.y_axis_label = config.y_axis_label
        self.smoothing_dict = config.smoothing_dict
        self.x_range = config.x_range
        self.y_range = config.y_range
        self.scale_intensity = config.scale_intensity

    @abstractmethod
    def plot(self, transitionGroup: TransitionGroup, features: Optional[List[TransitionGroupFeature]] = None, plot_type: Literal['chromatogram', 'mobilogram', 'spectrum'] = 'chromatogram'):
        pass




