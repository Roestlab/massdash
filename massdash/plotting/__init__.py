"""
massdash/plotting
~~~~~~~~~~~~~~~~~

This subpackage contains classes for plotting 
"""

from .InteractivePlotter import InteractivePlotter
from .InteractiveThreeDimensionPlotter import InteractiveThreeDimensionPlotter
from .InteractiveTwoDimensionPlotter import InteractiveTwoDimensionPlotter
from .GenericPlotter import PlotConfig, GenericPlotter

__all__ = ["InteractivePlotter", "InteractiveThreeDimensionPlotter", "InteractiveTwoDimensionPlotter", "PlotConfig", "GenericPlotter"]