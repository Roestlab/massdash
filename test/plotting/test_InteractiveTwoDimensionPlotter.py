from pathlib import Path
import pytest
import pandas as pd
import numpy as np
from bokeh.models import CrosshairTool


from massdash.testing.NumpySnapshotExtension import NumpySnapshotExtenstion
from massdash.testing.BokehSnapshotExtension import BokehSnapshotExtension
from massdash.plotting.InteractiveTwoDimensionPlotter import InteractiveTwoDimensionPlotter
from massdash.structs.FeatureMap import FeatureMap
from massdash.plotting.GenericPlotter import PlotConfig
from massdash.util import find_git_directory

TEST_PATH = find_git_directory(Path(__file__).resolve()).parent / 'test' / 'test_data'

@pytest.fixture
def snapshot_bokeh(snapshot):
    return snapshot.use_extension(BokehSnapshotExtension)

@pytest.fixture
def snapshot_numpy(snapshot):
    return snapshot.use_extension(NumpySnapshotExtenstion)

@pytest.fixture
def featureMap():
    df = pd.read_csv(TEST_PATH / 'featureMap' / 'ionMobilityTestFeatureDf.tsv', sep='\t')
    return FeatureMap(df, sequence='TEST', precursor_charge=2)

'''
@pytest.fixture(params=[
    dict(include_ms1=True, include_ms2=True, title="Title", subtitle="Subtitle", smoothing_type=dict(type='none'), type_of_heatmap='m/z vs retention time'),
    dict(include_ms1=True, include_ms2=False, title=None, subtitle=None, smoothing_type=dict(type='none'), type_of_heatmap='m/z vs ion mobility'),
    dict(include_ms1=True, include_ms2=False, title=None, subtitle=None, smoothing_type=dict(type='none'), type_of_heatmap='m/z vs ion mobility'),
    dict(include_ms1=True, include_ms2=True, title="Title", subtitle="Title", smoothing_type=dict(type='none'), scale_intensity=True),
    dict(include_ms1=True, include_ms2=True, title="Title", subtitle="Title", smoothing_type=dict(type='sgolay', sgolay_polynomial_order=3, sgolay_frame_length=5), scale_intensity=True),
    dict(include_ms1=True, include_ms2=True, title="Title", subtitle="Title", smoothing_type=dict(type='gauss', gaussian_sigma=2.0, gaussian_window=5), scale_intensity=True),
])
def genericPlotter(request):
    config_dict = dict(include_ms1=True, 
                       include_ms2=True, 
                       title="Title", 
                       subtitle="Subtitle", 
                       type_of_heatmap='retention time vs ion mobility')
    config = PlotConfig()
    config.update(request.param)
    return InteractiveTwoDimensionPlotter(plot_config=config)
'''

@pytest.fixture
def defaultPlotter():
    config = PlotConfig()
    return InteractiveTwoDimensionPlotter(config)

@pytest.mark.parametrize('heatmap,expected1,expected2', 
                         [('m/z vs retention time', 'Retention time (s)', 'm/z'), 
                          ('m/z vs ion mobility', 'Ion mobility (1/K0)', 'm/z'), 
                          ('retention time vs ion mobility', 'Retention time (s)', 'Ion mobility (1/K0)')])
def test_get_axis_titles(heatmap, expected1, expected2):
    config = PlotConfig()
    config.update(dict(type_of_heatmap=heatmap))
    plotter = InteractiveTwoDimensionPlotter(config)
    axis1, axis2 = plotter.get_axis_titles()
    assert axis1 == expected1
    assert axis2 == expected2
    
@pytest.mark.parametrize('index,columns', [('mz', 'rt'), ('mz', 'im'), ('im', 'rt')])
def test_get_plot_parameters(defaultPlotter, index, columns, featureMap, snapshot_numpy):
    arr = featureMap.feature_df.pivot_table(index=index, columns=columns, values='int', aggfunc="sum")
    expected_output = np.concatenate([np.array([i]) if isinstance(i, float) else i for i in defaultPlotter.get_plot_parameters(arr) ])

    assert snapshot_numpy == expected_output

@pytest.mark.parametrize('index,columns,smoothing_dict,normalization_dict', [
    # mz vs rt
    ('mz', 'rt', dict(type='none'), dict(type='none')),
    ('mz', 'rt', dict(type='gauss', sigma=1.2), dict(type='none')),
    ('mz', 'rt', dict(type='none'), dict(type='equalization', bins=2)),
    ('mz', 'rt', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # mz vs im
    ('mz', 'im', dict(type='none'), dict(type='none')),
    ('mz', 'im', dict(type='gauss', sigma=1.2), dict(type='none')),
    ('mz', 'im', dict(type='none'), dict(type='equalization', bins=2)),
    ('mz', 'im', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # im vs rt
    ('im', 'rt', dict(type='none'), dict(type='none')),
    ('im', 'rt', dict(type='gauss', sigma=1.2), dict(type='none')),
    ('im', 'rt', dict(type='none'), dict(type='equalization', bins=2)),
    ('im', 'rt', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2))])
def test_prepare_array(featureMap, index, columns, smoothing_dict, normalization_dict, snapshot_numpy):
    arr = featureMap.feature_df.pivot_table(index=index, columns=columns, values='int', aggfunc="sum")
    config = PlotConfig()
    config.update(dict(smoothing_dict=smoothing_dict, normalization_dict=normalization_dict, include_ms1=True, include_ms2=True))
    plotter = InteractiveTwoDimensionPlotter(config)
    output = np.array(plotter.prepare_array(arr))
    assert snapshot_numpy == output


def test_create_heatmap_plot(featureMap, defaultPlotter, snapshot_bokeh):
    arr = featureMap.feature_df.pivot_table(index='im', columns='rt', values='int', aggfunc="sum")

    im_arr, rt_arr, dw_main, dh_main, rt_min, rt_max, im_min, im_max = defaultPlotter.get_plot_parameters(arr) # TODO fill in manually do not rely on function
    linked_crosshair = CrosshairTool(dimensions="both")

    arr = arr.to_numpy()
    arr[np.isnan(arr)] = 0

    plot = defaultPlotter.create_heatmap_plot('title', arr, rt_min, rt_max, im_min, im_max, dw_main, dh_main, linked_crosshair, [])
    assert snapshot_bokeh == plot

@pytest.mark.parametrize('include_ms1,include_ms2,heatmap,smoothing_dict,normalization_dict', [
    ## include ms1 and ms2
    # mz vs rt
    (True, True, 'm/z vs retention time', dict(type='none'), dict(type='none')),
    (True, True, 'm/z vs retention time', dict(type='gauss', sigma=1.2), dict(type='none')),
    (True, True, 'm/z vs retention time', dict(type='none'), dict(type='equalization', bins=2)),
    (True, True, 'm/z vs retention time', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # mz vs im
    (True, True, 'm/z vs ion mobility', dict(type='none'), dict(type='none')),
    (True, True, 'm/z vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='none')),
    (True, True, 'm/z vs ion mobility', dict(type='none'), dict(type='equalization', bins=2)),
    (True, True, 'm/z vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # im vs rt
    (True, True, 'retention time vs ion mobility', dict(type='none'), dict(type='none')),
    (True, True, 'retention time vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='none')),
    (True, True, 'retention time vs ion mobility', dict(type='none'), dict(type='equalization', bins=2)),
    (True, True, 'retention time vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),

    ## Only MS1
    # mz vs rt
    (True, False, 'm/z vs retention time', dict(type='none'), dict(type='none')),
    (True, False, 'm/z vs retention time', dict(type='gauss', sigma=1.2), dict(type='none')),
    (True, False, 'm/z vs retention time', dict(type='none'), dict(type='equalization', bins=2)),
    (True, False, 'm/z vs retention time', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # mz vs im
    (True, False, 'm/z vs ion mobility', dict(type='none'), dict(type='none')),
    (True, False, 'm/z vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='none')),
    (True, False, 'm/z vs ion mobility', dict(type='none'), dict(type='equalization', bins=2)),
    (True, False, 'm/z vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # im vs rt
    (True, False, 'retention time vs ion mobility', dict(type='none'), dict(type='none')),
    (True, False, 'retention time vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='none')),
    (True, False, 'retention time vs ion mobility', dict(type='none'), dict(type='equalization', bins=2)),
    (True, False, 'retention time vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),

    ## Only MS2
    # mz vs rt
    (False, True, 'm/z vs retention time', dict(type='none'), dict(type='none')),
    (False, True, 'm/z vs retention time', dict(type='gauss', sigma=1.2), dict(type='none')),
    (False, True, 'm/z vs retention time', dict(type='none'), dict(type='equalization', bins=2)),
    (False, True, 'm/z vs retention time', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # mz vs im
    (False, True, 'm/z vs ion mobility', dict(type='none'), dict(type='none')),
    (False, True, 'm/z vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='none')),
    (False, True, 'm/z vs ion mobility', dict(type='none'), dict(type='equalization', bins=2)),
    (False, True, 'm/z vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # im vs rt
    (False, True, 'retention time vs ion mobility', dict(type='none'), dict(type='none')),
    (False, True, 'retention time vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='none')),
    (False, True, 'retention time vs ion mobility', dict(type='none'), dict(type='equalization', bins=2)),
    (False, True, 'retention time vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2))])
def test_plot_aggregated_heatmap(defaultPlotter, featureMap, include_ms1, include_ms2, heatmap, smoothing_dict, normalization_dict, snapshot_bokeh):
    config = PlotConfig()
    config.update(dict(smoothing_dict=smoothing_dict, normalization_dict=normalization_dict, include_ms1=include_ms1, include_ms2=include_ms2, type_of_heatmap=heatmap))
    plot = defaultPlotter.plot_aggregated_heatmap(featureMap)
    assert snapshot_bokeh == plot


@pytest.mark.parametrize('include_ms1,include_ms2,heatmap,smoothing_dict,normalization_dict', [
## include ms1 and ms2
    # mz vs rt
    (True, True, 'm/z vs retention time', dict(type='none'), dict(type='none')),
    (True, True, 'm/z vs retention time', dict(type='gauss', sigma=1.2), dict(type='none')),
    (True, True, 'm/z vs retention time', dict(type='none'), dict(type='equalization', bins=2)),
    (True, True, 'm/z vs retention time', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # mz vs im
    (True, True, 'm/z vs ion mobility', dict(type='none'), dict(type='none')),
    (True, True, 'm/z vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='none')),
    (True, True, 'm/z vs ion mobility', dict(type='none'), dict(type='equalization', bins=2)),
    (True, True, 'm/z vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # im vs rt
    (True, True, 'retention time vs ion mobility', dict(type='none'), dict(type='none')),
    (True, True, 'retention time vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='none')),
    (True, True, 'retention time vs ion mobility', dict(type='none'), dict(type='equalization', bins=2)),
    (True, True, 'retention time vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),

    ## Only MS1
    # mz vs rt
    (True, False, 'm/z vs retention time', dict(type='none'), dict(type='none')),
    (True, False, 'm/z vs retention time', dict(type='gauss', sigma=1.2), dict(type='none')),
    (True, False, 'm/z vs retention time', dict(type='none'), dict(type='equalization', bins=2)),
    (True, False, 'm/z vs retention time', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # mz vs im
    (True, False, 'm/z vs ion mobility', dict(type='none'), dict(type='none')),
    (True, False, 'm/z vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='none')),
    (True, False, 'm/z vs ion mobility', dict(type='none'), dict(type='equalization', bins=2)),
    (True, False, 'm/z vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # im vs rt
    (True, False, 'retention time vs ion mobility', dict(type='none'), dict(type='none')),
    (True, False, 'retention time vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='none')),
    (True, False, 'retention time vs ion mobility', dict(type='none'), dict(type='equalization', bins=2)),
    (True, False, 'retention time vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),

    ## Only MS2
    # mz vs rt
    (False, True, 'm/z vs retention time', dict(type='none'), dict(type='none')),
    (False, True, 'm/z vs retention time', dict(type='gauss', sigma=1.2), dict(type='none')),
    (False, True, 'm/z vs retention time', dict(type='none'), dict(type='equalization', bins=2)),
    (False, True, 'm/z vs retention time', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # mz vs im
    (False, True, 'm/z vs ion mobility', dict(type='none'), dict(type='none')),
    (False, True, 'm/z vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='none')),
    (False, True, 'm/z vs ion mobility', dict(type='none'), dict(type='equalization', bins=2)),
    (False, True, 'm/z vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2)),
    # im vs rt
    (False, True, 'retention time vs ion mobility', dict(type='none'), dict(type='none')),
    (False, True, 'retention time vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='none')),
    (False, True, 'retention time vs ion mobility', dict(type='none'), dict(type='equalization', bins=2)),
    (False, True, 'retention time vs ion mobility', dict(type='gauss', sigma=1.2), dict(type='equalization', bins=2))])
def test_plot_individual_heatmaps(featureMap, include_ms1, include_ms2, heatmap, smoothing_dict, normalization_dict, snapshot, snapshot_bokeh):
    config = PlotConfig()
    config.update(dict(include_ms1=include_ms1, include_ms2=include_ms2, smoothing_type=smoothing_dict, normalization_dict=normalization_dict, type_of_heatmap=heatmap, title="Title", subtitle="Subtitle"))
    plotter = InteractiveTwoDimensionPlotter(config)
    plots = plotter.plot_individual_heatmaps(featureMap)
    assert len(plots) == snapshot
    assert snapshot_bokeh == plots[0] # only check the first plot, checking list of plots not supported