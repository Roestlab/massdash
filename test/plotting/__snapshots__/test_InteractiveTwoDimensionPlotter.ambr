# serializer version: 1
# name: test_plot_individual_heatmaps[False-True-m/z vs ion mobility-smoothing_dict28-normalization_dict28]
  6
# ---
# name: test_plot_individual_heatmaps[False-True-m/z vs ion mobility-smoothing_dict29-normalization_dict29]
  6
# ---
# name: test_plot_individual_heatmaps[False-True-m/z vs ion mobility-smoothing_dict30-normalization_dict30]
  6
# ---
# name: test_plot_individual_heatmaps[False-True-m/z vs ion mobility-smoothing_dict31-normalization_dict31]
  6
# ---
# name: test_plot_individual_heatmaps[False-True-m/z vs retention time-smoothing_dict24-normalization_dict24]
  6
# ---
# name: test_plot_individual_heatmaps[False-True-m/z vs retention time-smoothing_dict25-normalization_dict25]
  6
# ---
# name: test_plot_individual_heatmaps[False-True-m/z vs retention time-smoothing_dict26-normalization_dict26]
  6
# ---
# name: test_plot_individual_heatmaps[False-True-m/z vs retention time-smoothing_dict27-normalization_dict27]
  6
# ---
# name: test_plot_individual_heatmaps[False-True-retention time vs ion mobility-smoothing_dict32-normalization_dict32]
  6
# ---
# name: test_plot_individual_heatmaps[False-True-retention time vs ion mobility-smoothing_dict33-normalization_dict33]
  6
# ---
# name: test_plot_individual_heatmaps[False-True-retention time vs ion mobility-smoothing_dict34-normalization_dict34]
  6
# ---
# name: test_plot_individual_heatmaps[False-True-retention time vs ion mobility-smoothing_dict35-normalization_dict35]
  6
# ---
# name: test_plot_individual_heatmaps[True-False-m/z vs ion mobility-smoothing_dict16-normalization_dict16]
  1
# ---
# name: test_plot_individual_heatmaps[True-False-m/z vs ion mobility-smoothing_dict17-normalization_dict17]
  1
# ---
# name: test_plot_individual_heatmaps[True-False-m/z vs ion mobility-smoothing_dict18-normalization_dict18]
  1
# ---
# name: test_plot_individual_heatmaps[True-False-m/z vs ion mobility-smoothing_dict19-normalization_dict19]
  1
# ---
# name: test_plot_individual_heatmaps[True-False-m/z vs retention time-smoothing_dict12-normalization_dict12]
  1
# ---
# name: test_plot_individual_heatmaps[True-False-m/z vs retention time-smoothing_dict13-normalization_dict13]
  1
# ---
# name: test_plot_individual_heatmaps[True-False-m/z vs retention time-smoothing_dict14-normalization_dict14]
  1
# ---
# name: test_plot_individual_heatmaps[True-False-m/z vs retention time-smoothing_dict15-normalization_dict15]
  1
# ---
# name: test_plot_individual_heatmaps[True-False-retention time vs ion mobility-smoothing_dict20-normalization_dict20]
  1
# ---
# name: test_plot_individual_heatmaps[True-False-retention time vs ion mobility-smoothing_dict21-normalization_dict21]
  1
# ---
# name: test_plot_individual_heatmaps[True-False-retention time vs ion mobility-smoothing_dict22-normalization_dict22]
  1
# ---
# name: test_plot_individual_heatmaps[True-False-retention time vs ion mobility-smoothing_dict23-normalization_dict23]
  1
# ---
# name: test_plot_individual_heatmaps[True-True-m/z vs ion mobility-smoothing_dict4-normalization_dict4]
  7
# ---
# name: test_plot_individual_heatmaps[True-True-m/z vs ion mobility-smoothing_dict5-normalization_dict5]
  7
# ---
# name: test_plot_individual_heatmaps[True-True-m/z vs ion mobility-smoothing_dict6-normalization_dict6]
  7
# ---
# name: test_plot_individual_heatmaps[True-True-m/z vs ion mobility-smoothing_dict7-normalization_dict7]
  7
# ---
# name: test_plot_individual_heatmaps[True-True-m/z vs retention time-smoothing_dict0-normalization_dict0]
  7
# ---
# name: test_plot_individual_heatmaps[True-True-m/z vs retention time-smoothing_dict1-normalization_dict1]
  7
# ---
# name: test_plot_individual_heatmaps[True-True-m/z vs retention time-smoothing_dict2-normalization_dict2]
  7
# ---
# name: test_plot_individual_heatmaps[True-True-m/z vs retention time-smoothing_dict3-normalization_dict3]
  7
# ---
# name: test_plot_individual_heatmaps[True-True-retention time vs ion mobility-smoothing_dict10-normalization_dict10]
  7
# ---
# name: test_plot_individual_heatmaps[True-True-retention time vs ion mobility-smoothing_dict11-normalization_dict11]
  7
# ---
# name: test_plot_individual_heatmaps[True-True-retention time vs ion mobility-smoothing_dict8-normalization_dict8]
  7
# ---
# name: test_plot_individual_heatmaps[True-True-retention time vs ion mobility-smoothing_dict9-normalization_dict9]
  7
# ---
